File: 004_FIB/004_FIB.R
if(FALSE){
  "
  Title: Rabbit and Recurrence Relations
  Rosalind ID: FIB
  Rosalind #: 004
  URL: http://rosalind.info/problems/fib

  Concept - Rabbits reproducing k offspring each month while
  you assume immortality.  Return how many rabbits after 'n'
  months.
  "
}

if(FALSE){
}

fib = function(n,k){
  a <- c(1,1)  
  for(i in seq(n-2)){  
    if(i == 1){
      a = append(a, a[2]+(a[1]*k))
    }else{
      a[1] = a[2]
      a[2] = a[3]
      a[3] = a[3]+(a[1]*k)
    }
  }
  return(a)
}

fib_list = function(n,k){
  fib_table <- c(1)
  for(i in seq(n-1)){
    if(i < 2){
      fib_table = append(fib_table, 1)
    }else{
      fib_table = append(fib_table, (fib_table[length(fib_table)]+(fib_table[length(fib_table)-1]*k)))
    }
  }
  return(fib_table)
}

File: 004_FIB/004_FIB.py
'''
Title: Rabbit and Recurrence Relations
Rosalind ID: FIB
Rosalind #: 004
URL: http://rosalind.info/problems/fib
'''

def rabbit_pairs(n, k):
    pairs = [0] * n
    pairs[0] = 1

    # Calculate the number of rabbit pairs for each subsequent month using the recurrence relation
    for i in range(1, n):
        pairs[i] = pairs[i - 1] + k * (pairs[i - 2] if i >= 2 else 0)

    return pairs[-1]

n, k = map(int, input().split())
result = rabbit_pairs(n, k)
print(result)

